### ISSUE TITLE
gentle_asarray structured array issues

### ISSUE THREAD

**Issue by braingram** [spacetelescope/stdatamodels#188]
While looking at updating [gentle_asarray](https://github.com/spacetelescope/stdatamodels/blob/078bd1c504482f6d09118d2d240c449e4c71f99f/src/stdatamodels/util.py#L24) I noticed it does some odd things depending on the input.

1) Modification of provided dtype:
```python
a = np.zeros(2, dtype=[('f', 'f8'), ('i', 'i4')])
ndt = np.dtype([('F', 'f8'), ('i', 'i4')])  # capitalized 'F'
b = gentle_asarray(a, ndt)
assert b.dtype == ndt
assert ndt == np.dtype([('f', 'f8'), ('i', 'i4')]) # ndt no longer has a capital 'F'
```

2) swapping column names but not subdtypes:
```python
a = np.zeros(2, dtype=[('f', 'f8'), ('i', 'i4')])
ndt = np.dtype([('i', 'i4'), ('f', 'f8')])  # swap 'i' and 'f'
b = gentle_asarray(a, ndt)
assert b.dtype == ndt
assert ndt == np.dtype([('f', '<i4'), ('i', '<f8')]) # note that 'f' now has a 'i4' subdtype
```

Looking at the usage of `gentle_asarray` I only see one use in `properties`: https://github.com/spacetelescope/stdatamodels/blob/078bd1c504482f6d09118d2d240c449e4c71f99f/src/stdatamodels/properties.py#L81-L85
as this use generate a new dtype prior to the call and doesn't use the dtype (beyond the `fields` access shown above) after the function the modification of the provided dtype seems innocuous (although I think this could be fixed by copying the dtype). The swapping of the column names but not subdtypes seems more problematic. One possible issue can be seen with `NirspecflatModel`:
```python
m = dm.NirspecFlatModel()
assert m.flat_table.dtype == np.dtype([('slit_name', 'S15'), ('nelem', '<i4'), ('wavelength', 'f4', (130000,)), ('data', 'f4', (130000,))])
# make a 'table' where 'nelem' and 'slit_name' order is swapped
b = np.zeros(2, dtype=[('nelem', '<i4'), ('slit_name', 'S15'), ('wavelength', 'f4', (130000,)), ('data', 'f4', (130000,))])
b['nelem'] = 2
b['slit_name'] = '1'
m.flat_table = b
# note the odd type which does not match the schema
assert m.flat_table.dtype == np.dtype([('nelem', 'S15'), ('slit_name', '<i4'), ('wavelength', 'f4', (130000,)), ('data', 'f4', (130000,))])
assert m.flat_table['nelem'][0] == '2'  # note that nelem is defined in the schema as 'i4' but here is a 'S15'
m.save('foo.fits')  # saving the model causes no error
m = dm.open('foo.fits')  # nor does opening the saved model
assert m.flat_table['nelem'][0] == '2'  # which still has the incorrect table
```

**Comment by braingram** [comment]
I think issue 1 above (the modification of the provided dtype) is allowing the `test_gentle_asarray_field_name_case` test to pass even though the name conversion is not performed.
https://github.com/spacetelescope/stdatamodels/blob/078bd1c504482f6d09118d2d240c449e4c71f99f/tests/test_util.py#L133-L142
Specifically, I think the expected dtype should have names 'COL1' and 'COL2' but the resulting array has 'col1' and 'col2' and the test on main passes only because the passed in dtype is modified.

**Comment by braingram** [comment]
Related to this, the current MIRI cubepars reference file contains a table that does not match the schema.
https://jwst-crds.stsci.edu/browse/jwst_miri_cubepar_0006.fits
The 'CUBEPAR' table read from the fits file has the following dtype:
```dtype((numpy.record, [('CHANNEL', '>i2'), ('BAND', 'S12'), ('WAVEMIN', '>f4'), ('WAVEMAX', '>f4'), ('SPAXELSIZE', '>f4'), ('SPECTRALSTEP', '>f8')]))```
whereas the schema defines a different order and a few different subdtypes:
https://github.com/spacetelescope/stdatamodels/blob/078bd1c504482f6d09118d2d240c449e4c71f99f/src/stdatamodels/jwst/datamodels/schemas/miri_ifucubepars.schema.yaml#L10-L25
When the datamodel is opened, currently `gentle_asarray` will end up producing a third dtype:
```python
>> m = dm.open('jwst_miri_cubepar_0006.fits')
>> m.ifucubepars_table.dtype
dtype([('CHANNEL', 'S3'), ('BAND', 'S12'), ('WAVEMIN', '<f4'), ('WAVEMAX', '<f4'), ('SPAXELSIZE', '<f4'), ('SPECTRALSTEP', '<f4')])
```
To summarize, 3 dtypes get used
| In file                                  | In schema                | Opened datamodel  |
| ------------------------- | ------------------- | -------------------- |
| CHANNEL, i2                     | channel, **s3**       | CHANNEL, s3           |
| BAND, s12                         | band, s12                  | BAND, s12                |
| WAVEMIN, f4                    | **spaxelsize**, f4      | WAVEMIN, f4           |
| WAVEMAX, f4                   | **spectralstep**, f4  | WAVEMAX, f4          |
| SPAXELSIZE, f4                | **wavemin**, f4        | SPAXELSIZE, f4       |
| SPECTRALSTEP, **f8**   | **wavemax**, f4        | SPECTRALSTEP, f4 |

To illustrate this issue, if I modify the `spectral step` schema definition to be `f8` the datamodel produced from opening the file has `WAVEMAX` as `f8`:
```dtype([('CHANNEL', 'S3'), ('BAND', 'S12'), ('WAVEMIN', '<f4'), ('WAVEMAX', '<f8'), ('SPAXELSIZE', '<f4'), ('SPECTRALSTEP', '<f4')])```

**Comment by braingram** [comment]
The apcorr reference file presents a slightly different issue.
https://jwst-crds.stsci.edu/browse/jwst_nircam_apcorr_0004.fits
This contains a table with a dtype that contains big-endian data:
```dtype((numpy.record, [('filter', 'S12'), ('pupil', 'S15'), ('eefraction', '>f4'), ('radius', '>f4'), ('apcorr', '>f4'), ('skyin', '>f4'), ('skyout', '>f4')]))```
On my system the schema dtype ends up with default little endian. The fits table behaves very oddly when taking a view with a different endianess. For example:
```python
ff = fits.open('jwst_nircam_apcorr_0004.fits')
example = ff['APCORR'].data[:2]  # just use the first two rows for this example
v = example.view(dtype=[('filter', 'S12'), ('pupil', 'S15'), ('eefraction', '<f4'), ('radius', '<f4'), ('apcorr', '<f4'), ('skyin', '<f4'), ('skyout', '<f4')])  # make a view with little endian data
assert v['radius'][0] == example['radius'][0]   # data in view is ok
mask = [True, False]  # now mask out the second row
masked_example = example[mask]
masked_view = v[mask]
assert masked_view['radius'][0] != masked_example['radius'][0]  # after masking the data do not agree
assert abs(masked_example['radius'][0] - 0.451) < 0.001
assert masked_view['radius'][0] > 1E20  # value after masking is completely wrong
```

**Comment by braingram** [comment]
Fixed in #189
